zzzz_modular_occulus/code/game/machinery/dorms_elevator.dm

```diff
@@ -0,0 +1,131 @@
+
+	//When the passenger is put into storage, their respawn time is reduced.
+	//This check exists for the benefit of people who get put into cryostorage while SSD and come back later
+	if (passenger.in_perfect_health())
+		if (passenger.mind && passenger.mind.key)
+
+			//Whoever inhabited this body is long gone, we need some black magic to find where and who they are now
+			var/mob/M = key2mob(passenger.mind.key)
+			if (istype(M))
+				if (!(M.get_respawn_bonus("CRYOSLEEP")))
+					//We send a message to the passenger's current mob - probably a ghost, but who knows.
+					to_chat(M, SPAN_NOTICE("Because you returned to the dorms, your crew respawn time has been reduced by [CRYOPOD_SPAWN_BONUS_DESC]."))
+					M << 'sound/effects/magic/blind.ogg' //Play this sound to a player whenever their respawn time gets reduced
+
+				//Going safely to cryo will allow the patient to respawn more quickly
+				M.set_respawn_bonus("CRYOSLEEP", CRYOPOD_SPAWN_BONUS)
+
+	// This despawn is not a gib() in this sense, it is used to remove objectives tied on these despawned mobs in cryos
+	passenger.ghostize()
+	passenger.despawn()
+
+
+
+//// This code is for the buttons and airlock in the elevator in the dorms area that offers players
+//// an alternative method of leaving the round without having to cryo.
+
+/obj/machinery/button/remote/elevator_panel
+	name = "Elevator Panel"
+	desc = "An elevator panel used to return to the dorms once a shift is over."
+
+	var/cooldown = 24 SECONDS
+	var/cooldown_timer
+	var/list/preserve_items = list(
+		/obj/item/weapon/hand_tele,
+		/obj/item/weapon/card/id/captains_spare,
+		/obj/item/device/aicard,
+		/obj/item/device/mmi,
+		/obj/item/device/paicard,
+		/obj/item/weapon/gun,
+		/obj/item/weapon/pinpointer,
+		/obj/item/clothing/suit,
+		/obj/item/clothing/shoes/magboots,
+		/obj/item/blueprints,
+		/obj/item/clothing/head/space,
+		/obj/item/weapon/storage/internal
+	)
+	var/obj/item/device/radio/intercom/announce
+	var/on_store_message = "has clocked out of their shift and returned to the dormitories."
+	var/on_store_name = "Elevator Announcer"
+
+/obj/machinery/button/remote/elevator_panel/New()
+	announce = new /obj/item/device/radio/intercom(src)
+	update_icon()
+	..()
+
+/obj/machinery/button/remote/elevator_panel/trigger()
+	var/elevator_prompt
+	var/turf/T = get_turf(src)
+
+	if(world.time - cooldown_timer > cooldown)
+		cooldown_timer = world.time
+		elevator_prompt = alert(src, "Warning! Using the elevator will end your shift and remove you from the round. Are you -sure- you want to head to your dorm for the day?\n(Everyone inside the elevator with you will end their shift and be removed as well!)", "End your shift and head to the dorms?", "End shift", "Stay on shift")
+		if(elevator_prompt == "End shift")
+			for(var/obj/machinery/door/airlock/D in SSmachines.machinery)
```

Comment:
Maybe using an `in range(src, <somenum>)` (instead of `in SSmachines.machinery`) would be better here, instead of having to loop through every single door. Since (I presume) the elevator button will always be in close proximity to the airlock.

Reply:
Alternatively, just grab the area and check airlocks in that.

Reply:
The in range one was a really good idea. A lot less intensive. Got that in place too. Not committed yet.